### PR TITLE
BUG: fixed issue with mixed type groupby aggregate

### PR DESCRIPTION
**Pull request by nreeve17**
Fixes issue #16916, where using aggregate on a mixed type grouping
vector fails. Added test in test_aggregate.py to ensure that the bug
is fixed.

cc @ElDeveloper 

 - [x] closes #xxxx
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
@nreeve17 : Thanks for the PR!  I think you just need a `whatsnew`, and you should be set!

File: pandas/tests/groupby/test_aggregate.py [hunk 1]
```diff
@@ -892,3 +892,15 @@ def test_sum_uint64_overflow(self):
         expected.index.name = 0
         result = df.groupby(0).sum()
         tm.assert_frame_equal(result, expected)
+
+    def test_mixed_type_grouping(self):
```
Comment (gfyoung) [inline pandas/tests/groupby/test_aggregate.py]: Refer the issue number beneath the function definition.

File: pandas/core/groupby.py [hunk 2]
```diff
@@ -2883,7 +2883,12 @@ def aggregate(self, func_or_funcs, *args, **kwargs):
             except Exception:
                 result = self._aggregate_named(func_or_funcs, *args, **kwargs)
 
-            index = Index(sorted(result), name=self.grouper.names[0])
+            # mixed types fail to sort
+            try:
+                values = sorted(result)
+            except TypeError:
+                values = result
```
Comment (jreback) [inline pandas/core/groupby.py]: I would rather
```
# put at top
from pandas.core.agorithms

index = Index(safe_sort(result), name=self.grouper.names[0])
```

File: pandas/tests/groupby/test_aggregate.py [hunk 3]
```diff
@@ -892,3 +892,15 @@ def test_sum_uint64_overflow(self):
...
+        X = pd.DataFrame(data=[[[1, 1], [2, 2], [3, 3]],
```
Comment (jreback) [inline pandas/tests/groupby/test_aggregate.py]: expected =

File: pandas/tests/groupby/test_aggregate.py [hunk 4]
```diff
@@ -892,3 +892,15 @@ def test_sum_uint64_overflow(self):
...
+                               [[1, 1], [2, 2], [3, 3]]],
+                         columns=['X', 'Y', 'Z'],
+                         index=pd.Index(data=[2, 'g1'], name='grouping'))
+
+        S = pd.DataFrame(data=[[1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3]],
```
Comment (jreback) [inline pandas/tests/groupby/test_aggregate.py]: df =

File: pandas/tests/groupby/test_aggregate.py [hunk 5]
```diff
@@ -892,3 +892,15 @@ def test_sum_uint64_overflow(self):
...
+                         columns=list('XYZ'), index=list('qwer'))
+        S['grouping'] = ['g1', 'g1', 2, 2]
+        T = S.groupby('grouping').aggregate(lambda x: x.tolist())
```
Comment (jreback) [inline pandas/tests/groupby/test_aggregate.py]: result = df.groupby(....)

File: pandas/tests/groupby/test_aggregate.py [hunk 6]
```diff
@@ -892,3 +892,15 @@ def test_sum_uint64_overflow(self):
...
+        tm.assert_frame_equal(T, X)
```
Comment (jreback) [inline pandas/tests/groupby/test_aggregate.py]: tm.assert_frame_equal(result, expected)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby.py (hunk 2 above)
Comment (nreeve17) [inline pandas/core/groupby.py]: I tried this change and it caused 4 tests in the full test suite to fail. Those tests were TestDatetimeIndex.test_resample_categorical_data_with_timedeltaindex, TestGroupByAggregate.test_mixed_type_grouping (the test that I wrote), TestGroupBy.test_basic, and test_compressed_urls[python-explicit-gzip-.gz]
Also, I noticed that safe_sort returns all objects as strings, even if the input includes numeric types.

Comment (jreback) [inline pandas/core/groupby.py]: yeah this is an embedded issue with ``safe_sort``, fixed here: https://github.com/pandas-dev/pandas/pull/17034. you can rebase on top (or wait for me to merge this).

File: doc/source/whatsnew/v0.21.0.txt [hunk 7]
```diff
@@ -212,10 +212,11 @@ Plotting
 Groupby/Resample/Rolling
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
-- Bug in ``DataFrame.resample(...).size()`` where an empty ``DataFrame`` did not return a ``Series`` (:issue:`14962`)
-- Bug in :func:`infer_freq` causing indices with 2-day gaps during the working week to be wrongly inferred as business daily (:issue:`16624`)
-- Bug in ``.rolling(...).quantile()`` which incorrectly used different defaults than :func:`Series.quantile()` and :func:`DataFrame.quantile()` (:issue:`9413`, :issue:`16211`)
+- Bug in ``DataFrame.resample().size()`` where an empty ``DataFrame`` did not return a ``Series`` (:issue:`14962`)
+- Bug in ``infer_freq`` causing indices with 2-day gaps during the working week to be wrongly inferred as business daily (:issue:`16624`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: revert all of these changes except the this issue note

File: doc/source/whatsnew/v0.21.0.txt [hunk 8]
```diff
@@ -212,10 +212,11 @@ Plotting
...
+- Bug in ``.rolling.quantile()`` which incorrectly used different defaults than :func:`Series.quantile()` and :func:`DataFrame.quantile()` (:issue:`9413`, :issue:`16211`)
 - Bug in ``groupby.transform()`` that would coerce boolean dtypes back to float (:issue:`16875`)
+- Bug in ``Grouper.aggregate()`` on using a mixed type grouping vector (:issue:`16916`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: this is not at all clear, pls revise

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby.py (hunk 2 above)
Comment (nreeve17) [inline pandas/core/groupby.py]: I tried this and it resulted in the same tests failing, not sure what's going on.

Comment (jreback) [inline pandas/core/groupby.py]: did u rebase

Comment (nreeve17) [inline pandas/core/groupby.py]: I did 'git pull upstream master'

File: doc/source/whatsnew/v0.21.0.txt [hunk 9]
```diff
@@ -275,10 +275,11 @@ Plotting
 Groupby/Resample/Rolling
 ^^^^^^^^^^^^^^^^^^^^^^^^
 
+- Bug in ``DataFrame.resample().size()`` where an empty ``DataFrame`` did not return a ``Series`` (:issue:`14962`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: rebase error, pls remove

File: doc/source/whatsnew/v0.21.0.txt [hunk 10]
```diff
@@ -275,10 +275,11 @@ Plotting
...
 - Bug in ``DataFrame.resample(...).size()`` where an empty ``DataFrame`` did not return a ``Series`` (:issue:`14962`)
 - Bug in :func:`infer_freq` causing indices with 2-day gaps during the working week to be wrongly inferred as business daily (:issue:`16624`)
 - Bug in ``.rolling(...).quantile()`` which incorrectly used different defaults than :func:`Series.quantile()` and :func:`DataFrame.quantile()` (:issue:`9413`, :issue:`16211`)
-- Bug in ``groupby.transform()`` that would coerce boolean dtypes back to float (:issue:`16875`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: also a rebase error, you removed an existing issue

File: pandas/core/groupby.py (hunk 2 above)
Comment (jreback) [inline pandas/core/groupby.py]: you need to use ``safe_sort`` here with the latest changed.

``git pull upstream origin/master``

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by nreeve17** [comment]
I have re-added the bugfix line into the whatsnew doc, and I git pulled upstream master and pushed the changes with safe_sort.

**Comment by jreback** [comment]
looks like there is a ``TypeError`` raised somewhere which causes this to take an incorrect path.

**Comment by ElDeveloper** [comment]
@jreback, yeah, seems like the exception is coming from [this line](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/src/inference.pyx#L363).

**Comment by nreeve17** [comment]
@jreback Is there anything you might suggest to fix this error? This error did not show up when we used the Try/Catch block

**Comment by jreback** [comment]
you have to pick out the cases where its failing and see exactly how it is failing. its possible ``safe_sort`` needs another case

**Comment by jreback** [comment]
can you rebase / update

**Comment by nreeve17** [comment]
Commits should be updated 👍

**Comment by jreback** [comment]
can you rebase and update

**Review by jreback — CHANGES_REQUESTED** [review]

**Pull request closed without merging**